system/src/system_connection_manager.cpp

```diff
@@ -0,0 +1,500 @@
+        socklen_t len = sizeof(socketNetIfIndex);
+        sock_handle_t cloudSocket = system_cloud_get_socket_handle();
+        //int r = 
+        sock_getsockopt(cloudSocket, SOL_SOCKET, SO_BINDTODEVICE, &socketNetIfIndex, &len);
+        //LOG(TRACE, "getCloudConnectionNetwork %d : %lu", r, socketNetIfIndex);
+    }
+
+	return socketNetIfIndex;
+}
+
+network_handle_t ConnectionManager::selectCloudConnectionNetwork() {
+    network_handle_t bestNetwork = NETWORK_INTERFACE_ALL;
+
+    // 1: If there is a bound network connection. Do not use anything else, regardless of network state
+    network_handle_t boundNetwork;
+    size_t n = sizeof(boundNetwork);
+    // TODO: error handling
+    auto r = spark_get_connection_property(SPARK_CLOUD_BIND_NETWORK_INTERFACE, &boundNetwork, &n, nullptr);
+    LOG(TRACE, "%d selectCloudConnectionNetwork %lu", r, boundNetwork);
+
+    if (boundNetwork != NETWORK_INTERFACE_ALL) {
+        LOG(TRACE, "%d using bound network: %lu", r, boundNetwork);
+        return boundNetwork;
+    }
+
+    // 2: If no bound network, use preferred network
+    if (preferredNetwork_ != NETWORK_INTERFACE_ALL && spark::Network.from(preferredNetwork_).ready()) {
+        LOG(TRACE, "using preferred network: %lu", preferredNetwork_);
+        return preferredNetwork_;
+    }
+
+    // 3: If no preferred network, use the 'best' network based on criteria
+    // 3.1: Network is ready: ie configured + connected (see ipv4 routable hook)
+    // 3.2: Network has best criteria based on network tester stats (vector should be sorted in "best" order)
+    for (auto& i: *ConnectionTester::instance()->getConnectionMetrics()) {
+        if (spark::Network.from(i.interface).ready()) {
+            LOG(TRACE, "using best tested network: %lu", i.interface);
+            return i.interface;
+        }
+    }
+
+    // TODO: Determine a specific interface to bind to, even in the default case.
+    // ie: How should we handle selecting a cloud connection when no interfaces are up/ready?
+    // We should have some historical stats to rely on and then bring that network up? 
+    return bestNetwork;
+}
+
+ConnectionTester::ConnectionTester() {
+    network_interface_t interfaceList[] = { 
+        NETWORK_INTERFACE_ETHERNET,
+#if HAL_PLATFORM_WIFI 
+        NETWORK_INTERFACE_WIFI_STA, 
+#endif
+#if HAL_PLATFORM_CELLULAR
+        NETWORK_INTERFACE_CELLULAR
+#endif
+    };
+    
+    for (const auto& i: interfaceList) {
+        struct ConnectionMetrics interfaceDiagnostics = {};
+        interfaceDiagnostics.interface = i;
+        metrics_.append(interfaceDiagnostics);
+    }
+}
+
+ConnectionTester* ConnectionTester::instance() {
+    static ConnectionTester* tester = new ConnectionTester();
+    return tester;
+}
+
+static int getCloudHostnameAndPort(uint16_t * port, char * hostname, int hostnameLength) {
+    // Use this test server until real echo server is ready
+    strcpy(hostname, "publish-receiver-udp.particle.io");
```

Comment:
Should we revert this and use the actual server address even if the DTLS message isnt implemented yet? This means internet testing will fail until its deployed, but at least we wont be pointing at the publish receiver endpoint

Reply:
I think we have a new endpoint for this now?

Reply:
Yes this was removed